### PR TITLE
Changes to ThreadFactory and ExecutorFactory

### PR DESCRIPTION
**Pull request by kenfinnigan**
Make some adjustments in terms of how executors and threads are created and by what.

- Modify the JBoss `EnhancedQueueExecutor` to take a `ThreadFactory` from an optional build item. If one isn't present, create a `JBossThreadFactory` as previously done.
- Produce a `ThreadFactoryBuildItem` from Vert.x core extension containing a `ThreadFactory` producing `VertxThread` instances
- Modify the initialization of Vert.x to use the JBoss `EnhancedQueueExecutor` as the executor factory for any thread pools that Vert.x creates

### CODE REVIEW

**Comment by kenfinnigan** [comment]
Is the MongoDB Client integration test timeout something to worry about?

GIven it was the only failure, I'm loathed to rerun all the jobs just for that if it's a random error.

Anyone know?

File: docs/src/main/asciidoc/security-keycloak-authorization.adoc [hunk 1]
```diff
@@ -15,7 +15,7 @@ In other words, instead of explicitly enforcing access based on some specific ac
 
 By externalizing authorization from your application, you are allowed to protect your applications using different access control mechanisms as well as avoid re-deploying your application every time your security requirements change, where Keycloak will be acting as a centralized authorization service from where your protected resources and their associated permissions are managed.
 
-See the link:security-openid-connect[Using OpenID Connect to Protect Service Applications] guide for more information about `Bearer Token` authentication mechanism.
+See the link:security-openid-connect[Using OpenID Connect to Protect Service Applications] guide for more information about `Bearer Token` authentication mechanism. It is important to realize that it is the `Bearer Token` authentication mechanism which does the authentication and creates a security identity - while the `quarkus-keycloak-authorization` extension is responsible for applying a Keycloak Authorization Policy to this identity based on the current request path and other policy settings.
```
Comment (stuartwdouglas) [inline docs/src/main/asciidoc/security-keycloak-authorization.adoc:18]: Looks like this is supposed to be part of a different PR?

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java [hunk 2]
```diff
@@ -41,11 +43,17 @@
     @Override
     public CheckResult apply(RoutingContext routingContext, SecurityIdentity identity) {
 
+        if (identity.isAnonymous()) {
```
Comment (stuartwdouglas) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java:46]: Same here?

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+public class QuarkusExecutorFactory implements ExecutorServiceFactory {
+    private final VertxConfiguration conf;
+
+    public QuarkusExecutorFactory(VertxConfiguration conf) {
+        this.conf = conf;
+    }
+
+    @Override
+    public ExecutorService createExecutor(ThreadFactory threadFactory, Integer concurrency, Integer maxConcurrency) {
+        final EnhancedQueueExecutor.Builder builder = new EnhancedQueueExecutor.Builder()
```
Comment (stuartwdouglas) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: Can we just re-use the existing executor from ExecutorRecorder here?

My thinking is:
- Deprecate the config that sets the vert.x blocking thread pool size. If this is explicity set then use it, but print a warning
- Use Integer.MIN_VALUE to represent 'use the existing pool'. Pass this into the factory to tell it to just re-use the existing factory (this may get caught by validation, we may have to use MAX_VALUE instead)
- Allow the internal pool for file operations to be created as normal

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                } else {
+                    builder.setMaximumQueueSize(conf.queueSize.getAsInt());
+                }
+            }
+            builder.setGrowthResistance(conf.growthResistance);
+            builder.setKeepAliveTime(conf.keepAliveTime);
+        }
+
+        final EnhancedQueueExecutor eqe = builder.build();
+        eqe.prestartAllCoreThreads();
```
Comment (stuartwdouglas) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:44]: I don't think we want this for the internal pool, as the majority of applications will never use it. If you use the approach I mentioned above you can remove this.

**Review by stuartwdouglas — DISMISSED** [review]
Looks like there is an accidental Keycloak commit in here as well.

I think we should attempt to completely unify the pools, as we should only really need one. The fact that we have two at the moment is more of an internal implementation detail rather than a deliberate decision.

File: docs/src/main/asciidoc/security-keycloak-authorization.adoc (hunk 1 above)
Comment (kenfinnigan) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:18]: Didn't notice previously but I think my rebase went wonky.

Will rectify when I push updates

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 3 above)
Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: If the vert.x blocking thread pool size is set, by "use it" do you mean create a new EQE instance? Otherwise, I'm not sure how we can have the vert.x thread pool config override the overall quarkus thread pool config in _core_?

Right now all three pools in Vert.x use the same factory, so there's no way to separate out the regular blocking pool executor factory and shared executor pool, from the file operations one.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 4 above)
Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:44]: @dmlloyd had suggested we should to not impact performance.

If we really don't think that's a problem I can take it out.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 3 above)
Comment (dmlloyd) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: I guess you *could* return the same executor for all three.  That might introduce problems though if Vert.x is specifically designed to assume that certain thread pools never overlap.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 4 above)
Comment (stuartwdouglas) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:44]: I don't think it will matter much, in general perf is always significantly lower at startup as the JVM needs to warm up, I doubt the overhead of starting threads will be measurable in any meaningful way. The memory usage of the extra threads is measurable though, and there is a good chance that the vertx-internal pool will never be used by the majority of applications, as it is only used for 'async' file operations.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 3 above)
Comment (stuartwdouglas) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: I think you can merge the two main blocking ones, and leave the vertx-internal one as a separate pool. vertx-internal is only used for non blocking file operations (so super short tasks), and mostly will never be used.

The separation between the other two is kinda historical/based on technical limitations, there should be no issues with the vert.x pool being our main executor pool.

Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: `VertxImpl` creates three separate pools using the same executor factory:
- https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/VertxImpl.java#L157
- https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/VertxImpl.java#L160
- https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/VertxImpl.java#L1122

I don't think there's any way for us to know when we're creating an executor for blocking file operations vs anything else. At least without modifying Vert.x.

Are you saying we shouldn't be creating a separate `ExecutorFactory` for each call to the factory, but instead just pass out the EQE instance we create in Quarkus core? Essentially having a single thread pool for anything not running on the event loop for everything in Quarkus?

Comment (dmlloyd) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: > Are you saying we shouldn't be creating a separate `ExecutorFactory` for each call to the factory, but instead just pass out the EQE instance we create in Quarkus core? Essentially having a single thread pool for anything not running on the event loop for everything in Quarkus?

I was definitely *not* saying we *should* (and I'm not saying we *should not*), but I am saying that this is *possible*.

Another thing that is possible: they way they've implemented this means that we *could* keep a simple counter in our executor factory and know which pool is being created by its sequence.  We could decide which pool to create or return based on this information.  Ugly but effective.

Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: Apologies, I should've clarified I was directing that particular comment at @stuartwdouglas.

I'm fine with whatever approach we want to take, but I need to know which way to "jump" so I can finalize the PR and continue with the changes for OpenTelemetry

Comment (dmlloyd) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: Ah, sorry about that, I wasn't sure.

Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: Not a problem, and no need to be sorry.

Comment (stuartwdouglas) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/VertxImpl.java#L157 is the one that can be shared. We can make the factory know which one it is creating through the use of a 'magic value' (likely Integer.MAX_VALUE, as we can't easily pass in a negative number). 

Conceptually this pool and the ExecutorRecorder pool perform the same task, so I don't really see any reason to have the two pools.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java [hunk 5]
```diff
@@ -58,6 +58,29 @@
...
+    /**
+     * The executor growth resistance.
+     *
+     * A resistance factor applied after the core pool is full; values applied here will cause that fraction
+     * of submissions to create new threads when no idle thread is available. A value of {@code 0.0f} implies that
+     * threads beyond the core size should be created as aggressively as threads within it; a value of {@code 1.0f}
+     * implies that threads beyond the core size should never be created.
+     */
+    @ConfigItem
+    public float growthResistance;
```
Comment (gastaldi) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java:76]: No default value here?

File: core/runtime/src/main/java/io/quarkus/runtime/ExecutorRecorder.java [hunk 6]
```diff
@@ -123,9 +124,11 @@ public void run() {
         };
     }
 
-    private static EnhancedQueueExecutor createExecutor(ThreadPoolConfig threadPoolConfig) {
-        final JBossThreadFactory threadFactory = new JBossThreadFactory(new ThreadGroup("executor"), Boolean.TRUE, null,
-                "executor-thread-%t", JBossExecutors.loggingExceptionHandler("org.jboss.executor.uncaught"), null);
+    private static EnhancedQueueExecutor createExecutor(ThreadPoolConfig threadPoolConfig, ThreadFactory threadFactory) {
+        if (threadFactory == null) {
+            threadFactory = new JBossThreadFactory(new ThreadGroup("executor"), Boolean.TRUE, null,
```
Comment (FroMage) [inline core/runtime/src/main/java/io/quarkus/runtime/ExecutorRecorder.java:129]: So in which cases will this be `null` then?

I'm asking because for SR-CP I have a long-waiting branch that required JBossThreadFactory to let us create our own JBossThread subtype to have custom storage for ThreadLocals.

Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/ExecutorRecorder.java:129]: I guess it would be `null` if there is no Vert.x thread factory.  I assume the Vert.x thread factory also has the characteristics of the Netty thread factory BTW.

But I agree that the default should probably be to use the Netty factory or at least one with fast threadlocal support.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 3 above)
Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java:22]: @stuartwdouglas by using a magic number, do you mean setting Vert.x configuration for those pools that aren't shared to MAX_VALUE but then in the executor actually set it to the same value we have for the other pool config?

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java (hunk 5 above)
Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java:76]: I'd copied the properties across from our Thread configuration, which also didn't have defaults.

@dmlloyd should we include any defaults?

File: core/runtime/src/main/java/io/quarkus/runtime/ExecutorRecorder.java (hunk 6 above)
Comment (kenfinnigan) [reply core/runtime/src/main/java/io/quarkus/runtime/ExecutorRecorder.java:129]: Right. If the `vertx-core` extension is not present, then there will be no `ThreadFactory` present.

**Comment by kenfinnigan** [comment]
@stuartwdouglas for now I've backed out using a shared executor pool between Quarkus and Vert.x (though the shared one is still set on the factory for now), as it prevented applications from shutting down properly.

If you have some time to figure out why that's happening, that would be amazingly helpful

**Comment by kenfinnigan** [comment]
For those interested, you can see the broken shutdown by changing the start of https://github.com/quarkusio/quarkus/pull/17241/files#diff-84cabe6ae2e43ee2a6f8684f0b6be9cb8ceae3e7be530b662ec2b36d44f5aaeeR26 to do:
```
if (executorCount.incrementAndGet() == 1) {
  return sharedExecutor;
}
```

and then running the deployment tests in the vertx-http extension

**Comment by stuartwdouglas** [comment]
I have made an attempt at moving to a shared executor. There is a fair bit of complexity around this in dev mode, as the vert.x lifecycle does not exactly match the application lifecycle, as it needs to survive hot reload, but I think this should be ok.

**Comment by kenfinnigan** [comment]
Thanks very much @stuartwdouglas, appreciate you taking the time.

Fixed an issue with Vert.x Core test, hopefully it's all good now

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java (hunk 5 above)
Comment (dmlloyd) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java:76]: The default should be `0.0`.  Since this configuration property is primitive, the config processor is set up to make primitive properties have a default of their normal initializer value (in this case `0.0`).

Comment (kenfinnigan) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java:76]: So it's an implicit default and not one we need to add here?

Comment (dmlloyd) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/config/VertxConfiguration.java:76]: Right

**Comment by kenfinnigan** [comment]
I *think* I've fixed the nitpicky failures now, but there are a couple around DevMode I'm lost on at present.

@stuartwdouglas would you happen to have some time before your PTO to look into them?

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java (hunk 4 above)
Comment (johnaohara) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusExecutorFactory.java]: There were some tests where I was configuring the thread pool core threads `quarkus.thread-pool.core-threads` and that setting was being ignored. The behaviour was configurable `quarkus.thread-pool.prefill`: https://github.com/quarkusio/quarkus/pull/7599/files

**Comment by kenfinnigan** [comment]
FYI, I ended up changing tack a little because of my discoveries with the shared worker pool. As we're had separate worker pools between Quarkus and Vert.x to date, continuing that in DevMode only to circumvent thread issues I think is a pragmatic approach.

I also added configuration to prefill the executor pool if configuration is set, but it defaults to `false`

**Comment by kenfinnigan** [comment]
Would appreciate reviews on this today as I'd like to get it merged as a precursor to additional work for CR1

**Review by dmlloyd — APPROVED** [review]
